File: packages/python/elizaos/features/basic_capabilities/actions/reply.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from elizaos.generated.spec_helpers import require_action_spec
from elizaos.prompts import REPLY_TEMPLATE
from elizaos.types import Action, ActionExample, ActionResult, Content, ModelType
from elizaos.utils.spec_examples import convert_spec_examples
from elizaos.utils.xml import parse_key_value_xml

if TYPE_CHECKING:
    from elizaos.types import (
        HandlerCallback,
        HandlerOptions,
        IAgentRuntime,
        Memory,
        State,
    )

# Get text content from centralized specs
_spec = require_action_spec("REPLY")


def _convert_spec_examples() -> list[list[ActionExample]]:
    """Convert spec examples to ActionExample format."""
    return convert_spec_examples(_spec)


@dataclass
class ReplyAction:
    name: str = _spec["name"]
    similes: list[str] = field(default_factory=lambda: list(_spec.get("similes", [])))
    description: str = _spec["description"]

    async def validate(
        self, runtime: IAgentRuntime, _message: Memory, _state: State | None = None
    ) -> bool:
        return True

    async def handler(
        self,
        runtime: IAgentRuntime,
        message: Memory,
        state: State | None = None,
        options: HandlerOptions | None = None,
        callback: HandlerCallback | None = None,
        responses: list[Memory] | None = None,
    ) -> ActionResult:
        all_providers: list[str] = []
        if responses:
            for res in responses:
                if res.content and res.content.providers:
                    all_providers.extend(res.content.providers)

        state = await runtime.compose_state(
            message, [*all_providers, "RECENT_MESSAGES", "ACTION_STATE"]
        )

        template = REPLY_TEMPLATE
        if runtime.character.templates and "replyTemplate" in runtime.character.templates:
            template = runtime.character.templates["replyTemplate"]

        prompt = runtime.compose_prompt_from_state(state=state, template=template)

        response = await runtime.use_model(
            ModelType.TEXT_LARGE,
            {
                "prompt": prompt,
                "system": str(runtime.character.system or ""),
            },
        )

        parsed = parse_key_value_xml(response)
        thought = parsed.get("thought", "") if parsed else ""
        text = parsed.get("text", "") if parsed else ""

        thought = str(thought) if thought else ""
        text = str(text) if text else ""

        response_content = Content(
            thought=thought,
            text=text,
            actions=["REPLY"],
        )

        if callback:
            await callback(response_content)

        return ActionResult(
            text=f"Generated reply: {text}",
            values={
                "success": True,
                "responded": True,
                "lastReply": text,
                "lastReplyTime": runtime.get_current_time_ms(),
                "thoughtProcess": thought,
            },
            data={
                "actionName": "REPLY",
                "responseThought": thought,
                "responseText": text,
                "messageGenerated": True,
            },
            success=True,
        )

    @property
    def examples(self) -> list[list[ActionExample]]:
        return _convert_spec_examples()


reply_action = Action(
    name=ReplyAction.name,
    similes=ReplyAction().similes,
    description=ReplyAction.description,
    validate=ReplyAction().validate,
    handler=ReplyAction().handler,
    examples=ReplyAction().examples,
)

File: packages/typescript/src/features/basic-capabilities/actions/reply.ts
import { requireActionSpec } from "../../../generated/spec-helpers.ts";
import { logger } from "../../../logger.ts";
import { replyTemplate } from "../../../prompts.ts";
import type {
	Action,
	ActionExample,
	ActionResult,
	HandlerCallback,
	HandlerOptions,
	IAgentRuntime,
	Memory,
	State,
} from "../../../types/index.ts";
import { ModelType } from "../../../types/index.ts";
import { composePromptFromState, parseKeyValueXml } from "../../../utils.ts";

// Get text content from centralized specs
const spec = requireActionSpec("REPLY");

export const replyAction = {
	name: spec.name,
	similes: spec.similes ? [...spec.similes] : [],
	description: spec.description,
	validate: async (_runtime: IAgentRuntime) => {
		return true;
	},
	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		_options?: HandlerOptions,
		callback?: HandlerCallback,
		responses?: Memory[],
	): Promise<ActionResult> => {
		const actionContext = _options?.actionContext;
		const previousResults = actionContext?.previousResults || [];

		if (previousResults.length > 0) {
			logger.debug(
				{
					src: "plugin:basic-capabilities:action:reply",
					agentId: runtime.agentId,
					count: previousResults.length,
				},
				"Found previous action results",
			);
		}

		const allProviders: string[] = [];
		if (responses) {
			for (const res of responses) {
				const providers = res.content?.providers;
				if (providers && providers.length > 0) {
					allProviders.push(...providers);
				}
			}
		}

		state = await runtime.composeState(message, [
			...(allProviders ?? []),
			"RECENT_MESSAGES",
			"ACTION_STATE",
		]);

		const prompt = composePromptFromState({
			state,
			template: runtime.character.templates?.replyTemplate || replyTemplate,
		});

		const response = await runtime.useModel(ModelType.TEXT_LARGE, {
			prompt,
		});

		const parsedXml = parseKeyValueXml(response);
		const thoughtValue = parsedXml?.thought;
		const textValue = parsedXml?.text;
		const thought: string =
			typeof thoughtValue === "string" ? thoughtValue : "";
		const text: string = typeof textValue === "string" ? textValue : "";

		const responseContent = {
			thought,
			text,
			actions: ["REPLY"] as string[],
		};

		if (callback) {
			await callback(responseContent);
		}

		const now = Date.now();
		return {
			text: responseContent.text,
			values: {
				success: true,
				responded: true,
				lastReply: responseContent.text,
				lastReplyTime: now,
				thoughtProcess: thought,
			},
			data: {
				actionName: "REPLY",
				responseThought: thought,
				responseText: text,
				thought,
				messageGenerated: true,
			},
			success: true,
		};
	},
	examples: (spec.examples ?? []) as ActionExample[][],
} as Action;
